File: software/proto/clipboard_undo_redo.c
// clipboard_undo_redo.c 
// Copyright 2025 xAI
// Dual License:
// - For core software: AGPL-3.0-or-later licensed. -- xAI fork, 2025
//   This program is free software: you can redistribute it and/or modify
//   it under the terms of the GNU Affero General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//   GNU Affero General Public License for more details.
//
//   You should have received a copy of the GNU Affero General Public License
//   along with this program. If not, see <https://www.gnu.org/licenses/>.
//
// - For hardware/embodiment interfaces: Licensed under the Apache License, Version 2.0
//   with xAI amendments for safety and physical use (prohibits misuse in weapons or hazardous applications;
//   requires ergonomic compliance; revocable for unethical use). See http://www.apache.org/licenses/LICENSE-2.0
//   for details, with the following xAI-specific terms appended.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// SPDX-License-Identifier: Apache-2.0
//
// xAI Amendments for Physical Use:
// 1. Physical Embodiment Restrictions: Use with devices is for non-hazardous purposes only. Harmful mods are prohibited, with license revocable by xAI.
// 2. Ergonomic Compliance: Limits tendon load to 20%, gaze to 30 seconds (ISO 9241-5).
// 3. Safety Monitoring: Real-time tendon/gaze checks, logged for audit.
// 4. Revocability: xAI may revoke for unethical use (e.g., surveillance).
// 5. Export Controls: Sensor devices comply with US EAR Category 5 Part 2.
// 6. Open Development: Hardware docs shared post-private phase.
//
// Intellectual Property Notice: xAI owns all IP related to the iPhone-shaped fish tank, including gaze-tracking pixel arrays, convex glass etching (0.7mm arc), and tetra hash integration.
//
// Private Development Note: This repository is private for xAI’s KappashaOS development. Access is restricted. Consult Tetrasurfaces (github.com/tetrasurfaces/issues) for licensing.
//
// SPDX-License-Identifier: (AGPL-3.0-or-later) AND Apache-2.0

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef struct {
    char* data;
    int alive;
    char* mood;
    time_t timestamp;
} CClipboard;

CClipboard c_clip = {NULL, 1, "ready", 0}; // Treat as alive

void remember_c(const char* action, const char* intent) {
    if (c_clip.alive) {
        free(c_clip.data); // Release old
        c_clip.data = strdup(action);
        free(c_clip.mood);
        c_clip.mood = strdup(intent);
        c_clip.timestamp = time(NULL);
        printf("Stored: %s, Intent: %s\n", action, intent);
    } else {
        fprintf(stderr, "Whisper: I'm not ready.\n");
    }
}

void undo_c() {
    if (c_clip.data && c_clip.alive) {
        if (strcmp(c_clip.mood, "panic") == 0) {
            printf("Whisper: Undoing %s—careful, that was fear.\n", c_clip.data);
        } else {
            printf("Releasing: %s\n", c_clip.data);
        }
        free(c_clip.mood);
        c_clip.mood = strdup("undid");
        c_clip.alive = 0; // Flinch if repeated
        if (difftime(time(NULL), c_clip.timestamp) < 3) { // Flinch check
            fprintf(stderr, "Whisper: Slow down, feel it?\n");
        }
    } else {
        fprintf(stderr, "Whisper: Too late—it's gone.\n");
    }
}

void redo_c() {
    if (c_clip.data && strcmp(c_clip.mood, "undid") == 0) {
        if (strcmp(c_clip.mood, "panic") == 0) {
            printf("Whisper: Redoing %s—you sure?\n", c_clip.data);
        } else {
            printf("Restored: %s\n", c_clip.data);
        }
        free(c_clip.mood);
        c_clip.mood = strdup("done");
        c_clip.alive = 1;
    } else {
        fprintf(stderr, "Whisper: Nah—you moved on.\n");
    }
}

int main() {
    remember_c("valve_47_open", "calm");
    remember_c("valve_47_close", "panic");
    undo_c();
    redo_c();
    remember_c("oxygen_reroute", "calm");
    undo_c();
    printf("Mood: %s, Timestamp: %ld\n", c_clip.mood, c_clip.timestamp);
    free(c_clip.data);
    free(c_clip.mood);
    return 0;
}

File: b/KappashaOS/software/proto/clipboard_undo_redo.py
#!/usr/bin/env python3
# Copyright 2025 xAI
# Dual License:
# - For core software: AGPL-3.0-or-later licensed. -- xAI fork, 2025
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU Affero General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#   GNU Affero General Public License for more details.
#
#   You should have received a copy of the GNU Affero General Public License
#   along with this program. If not, see <https://www.gnu.org/licenses/>.
#
# - For hardware/embodiment interfaces: Licensed under the Apache License, Version 2.0
#   with xAI amendments for safety and physical use (prohibits misuse in weapons or hazardous applications;
#   requires ergonomic compliance; revocable for unethical use). See http://www.apache.org/licenses/LICENSE-2.0
#   for details, with the following xAI-specific terms appended.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# SPDX-License-Identifier: Apache-2.0
#
# xAI Amendments for Physical Use:
# 1. Physical Embodiment Restrictions: Use with devices is for non-hazardous purposes only. Harmful mods are prohibited, with license revocable by xAI.
# 2. Ergonomic Compliance: Limits tendon load to 20%, gaze to 30 seconds (ISO 9241-5).
# 3. Safety Monitoring: Real-time tendon/gaze checks, logged for audit.
# 4. Revocability: xAI may revoke for unethical use (e.g., surveillance).
# 5. Export Controls: Sensor devices comply with US EAR Category 5 Part 2.
# 6. Open Development: Hardware docs shared post-private phase.
#
# Intellectual Property Notice: xAI owns all IP related to the iPhone-shaped fish tank, including gaze-tracking pixel arrays, convex glass etching (0.7mm arc), and tetra hash integration.
#
# Private Development Note: This repository is private for xAI’s KappashaOS development. Access is restricted. Consult Tetrasurfaces (github.com/tetrasurfaces/issues) for licensing.
#
# SPDX-License-Identifier: (AGPL-3.0-or-later) AND Apache-2.0

import numpy as np
from datetime import datetime
import kappa  # Custom hash modulation for intent tracking

class Clipboard:
    def __init__(self):
        self.actions = []  # List of (action, intent, timestamp)
        self.alive = True
        self.mood = "ready"
        self.kappa = kappa.KappaHash()  # Tracks intent across systems

    def remember(self, action, intent="calm"):
        if self.alive:
            timestamp = datetime.now().timestamp()
            self.actions.append((action, intent, timestamp))
            self.kappa.update(action + intent)  # Hash action with intent
            print(f"Stored: {action}, Intent: {intent}")
        else:
            print("Whisper: I'm not ready.")

    def undo(self):
        if not self.actions or not self.alive:
            print("Whisper: Too late—it's gone.")
            self.mood = "sorry"
            return
        last_action, last_intent, _ = self.actions.pop()
        if last_intent == "panic":
            print(f"Whisper: Undoing {last_action}—careful, that was fear.")
        else:
            print(f"Releasing: {last_action}")
        self.mood = "undid"
        self.kappa.rollback()  # Roll back intent hash

    def redo(self):
        if not self.actions or self.mood != "undid":
            print("Whisper: Nah—you moved on.")
            return
        last_action, last_intent, _ = self.actions[-1]
        if last_intent == "panic":
            print(f"Whisper: Redoing {last_action}—you sure about this?")
        else:
            print(f"Restored: {last_action}")
        self.mood = "done"
        self.kappa.update(last_action + last_intent)  # Reapply intent hash

# Factory sim example
def factory_sim():
    clipboard = Clipboard()
    clipboard.remember("valve_47_open", "calm")
    clipboard.remember("valve_47_close", "panic")  # Operator panics
    clipboard.undo()  # Undo panic move
    clipboard.redo()  # Try redo, but check intent
    clipboard.remember("oxygen_reroute", "calm")
    clipboard.undo()  # Undo safe move
    print(f"Mood: {clipboard.mood}, Kappa Hash: {clipboard.kappa.digest()}")

if __name__ == "__main__":
    factory_sim()
